client/fingerprint/network_test.go

```diff
@@ -269,54 +277,78 @@ func TestNetworkFingerprint_default_device_absent(t *testing.T) {
 		Attributes: make(map[string]string),
 	}
-	cfg := &config.Config{NetworkSpeed: 100, NetworkInterface: "lo"}
 
-	request := &FingerprintRequest{Config: cfg, Node: node}
-	var response FingerprintResponse
-	err := f.Fingerprint(request, &response)
-	if err != nil {
-		t.Fatalf("err: %v", err)
+	testCases := []struct {
+		name         string
+		config       *config.Config
+		expectedCIDR string
+	}{
+		{
+			name:         "Loopback IPv6",
+			config:       &config.Config{NetworkSpeed: 100, NetworkInterface: "lo", PreferredAddressFamily: "ipv6"},
+			expectedCIDR: loCIDRv6,
+		},
+		{
+			name:         "Loopback IPv4",
+			config:       &config.Config{NetworkSpeed: 100, NetworkInterface: "lo", PreferredAddressFamily: "ipv4"},
+			expectedCIDR: loCIDRv4,
+		},
+		{
+			name:         "Loopback No preferred address family",
+			config:       &config.Config{NetworkSpeed: 100, NetworkInterface: "lo"},
+			expectedCIDR: loCIDRv6,
+		},
 	}
 
-	if !response.Detected {
-		t.Fatalf("expected response to be applicable")
-	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			request := &FingerprintRequest{Config: tc.config, Node: node}
+			var response FingerprintResponse
+			err := f.Fingerprint(request, &response)
+			if err != nil {
+				t.Fatalf("err: %v", err)
+			}
 
-	attributes := response.Attributes
-	if len(attributes) == 0 {
-		t.Fatalf("should apply")
-	}
+			if !response.Detected {
+				t.Fatalf("expected response to be applicable")
+			}
 
-	assertNodeAttributeContains(t, attributes, "unique.network.ip-address")
+			attributes := response.Attributes
+			if len(attributes) == 0 {
+				t.Fatalf("should apply")
+			}
 
-	ip := attributes["unique.network.ip-address"]
-	match := net.ParseIP(ip)
-	if match == nil {
-		t.Fatalf("Bad IP match: %s", ip)
-	}
+			assertNodeAttributeContains(t, attributes, "unique.network.ip-address")
 
-	if len(response.NodeResources.Networks) == 0 {
-		t.Fatal("Expected to find Network Resources")
-	}
+			ip := attributes["unique.network.ip-address"]
+			match := net.ParseIP(ip)
+			if match == nil {
+				t.Fatalf("Bad IP match: %s", ip)
+			}
 
-	// Test at least the first Network Resource
-	net := response.NodeResources.Networks[0]
-	if net.IP == "" {
-		t.Fatal("Expected Network Resource to not be empty")
-	}
-	if net.CIDR == "" {
-		t.Fatal("Expected Network Resource to have a CIDR")
-	}
-	if net.Device == "" {
-		t.Fatal("Expected Network Resource to have a Device Name")
-	}
-	if net.MBits == 0 {
-		t.Fatal("Expected Network Resource to have a non-zero bandwidth")
+			if len(response.NodeResources.Networks) == 0 {
+				t.Fatal("Expected to find Network Resources")
+			}
+
+			// Test at least the first Network Resource
+			net := response.NodeResources.Networks[0]
+			if net.IP == "" {
+				t.Fatal("Expected Network Resource to not be empty")
+			}
+			if strings.ToUpper(net.CIDR) == strings.ToUpper(tc.expectedCIDR) {
+				t.Errorf("Expected CIDR %s; got %s", tc.expectedCIDR, net.IP)
```

Comment:
Nice catch!
I'm a bit confused about how CIDR `127.0.0.1/8` result in IP address `127.0.0.0` instead of `127.0.0.1`

I'm going to compare the IP instead of the CIDR. In the end, we want to make sure they are on the expected family.

Reply:
looks like `127.0.0.0` is coming from the test `NetworkInterfaceDetectorOnlyLo`'s usage of [`net.ParseCIDR`](https://pkg.go.dev/net#ParseCIDR)

> It returns the IP address and the network implied by the IP and prefix length. For example, ParseCIDR("192.0.2.1/24") returns the IP address 192.0.2.1 and the network 192.0.2.0/24.

the mockish test interface implementation is only using the `IPNet` and discarding `IP`.

I agree using the IP for this test is fine, since it does confirm the by-family sorting behavior we're expecting to see. :+1: